### ISSUE TITLE
Using formatResults on a beforeFind crashes

### ISSUE THREAD

**Issue by elboletaire** [cakephp/cakephp#9787]
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: > 3.3.x. Checked under 3.3.8 and 3.3.9
* Platform and Target: All. Checked under linux, but should fail in all systems.

### What you did
Use `formatResults` on a model's or behavior's `beforeFind`.

I've created an empty project where you can see the error: https://github.com/elboletaire/matching-issue.

@lorenzo said he was able to reproduce this error under cakephp's test suite. I'm a bit lost on this, but if you tell me those test lines I could try to do a PR during this week.

### What happened
The app crashes with the error 

    Argument 1 passed to mapEntity() must be an instance of Cake\ORM\Entity, array given, called in {class} on line 65

\* Error reduced. In my project example crashes due to a `beforeFind` calling a finder with `formatResults` on a behavior.

### What you expected to happen
Execute `formatResults` without errors on a `beforeFind`, the way it works with a normal Query.

**Comment by elboletaire** [comment]
Obviously the bot has not properly parsed the version number :_D

**Comment by lorenzo** [comment]
Here's a failing test case. The culprit of the error is adding the result formatter on `beforeFind`.

```php
    /**
     * Tests that formatters cna be applied to deep associations that are fetched using
     * additional queries
     *
     * @return void
     */
    public function testFormatDeepDistantAssociationRecords2()
    {
        $table = TableRegistry::get('authors');
        $table->hasMany('articles');
        $articles = $table->association('articles')->target();
        $articles->hasMany('articlesTags');
        $tags = $articles->association('articlesTags')->target()->belongsTo('tags');

        $tags->target()->eventManager()->on('Model.beforeFind', function ($e, $query) {
            return $query->formatResults(function ($results) {
                return $results->map(function (Entity $tag) {
                    $tag->name .= ' - visited';

                    return $tag;
                });
            });
        });

        $query = $table->find()->contain(['articles.articlesTags.tags']);

        $query->mapReduce(function ($row, $key, $mr) {
            foreach ((array)$row->articles as $article) {
                foreach ((array)$article->articles_tags as $articleTag) {
                    $mr->emit($articleTag->tag->name);
                }
            }
        });

        $expected = ['tag1 - visited', 'tag2 - visited', 'tag1 - visited', 'tag3 - visited'];
        $this->assertEquals($expected, $query->toArray());
    }
```

**Comment by elboletaire** [comment]
:+1:

**Comment by elboletaire** [comment]
Yup, this does not only happen with the `beforeFind`. It does also happen due to the `matching`. In the project example, if I remove the `beforeFind` from the Behavior and attach the custom finder directly to the query (instead of by using the `beforeFind`) the app crashes the same way.

If i comment the `matching` the `formatResults` works as expected (does not crash).

**Comment by elboletaire** [comment]
But I'm not able to reproduce it with tests >_<

**Comment by elboletaire** [comment]
Why is this `onHold`? 😢

**Comment by lorenzo** [comment]
It is not on hold, just that no one has had the time to look into it.

**Comment by lorenzo** [comment]
@elboletaire maybe you feel brave to try fixing it?

**Comment by lorenzo** [comment]
@elboletaire It seems like this bug was fixed in 3.4. I'll push the test case above to the repo so that we don't have a regression.

**Comment by elboletaire** [comment]
uo! I'll give it a try then after upgrading (I downgraded to 3.3 again due to #10219)
